### ISSUE TITLE
DocumentReference from QueryDocumentSnapshot in model converter not typed correctly.

### ISSUE THREAD

**Issue by letsgamedotdev** [firebase/firebase-js-sdk#4278]
### [REQUIRED] Describe your environment

  * Operating System version: Windows
  * Browser version: Chrome  (87.0.4280.141)
  * Firebase SDK version: 9.1.0
  * Firebase Product: firestore

### [REQUIRED] Describe the problem

When trying to save the DocumentReference from the QueryDocumentSnapshot in the ModelConverter.fromFirestore function. The resulting document reference object is missing several fields.

#### Steps to reproduce:

See code below:

#### Relevant Code:

```
interface MyModel {
  ref: firebase.firestore.DocumentReference;
}

const modelConverter = {
  toFirestore: function (model: MyModel) {
    return _.omit(model, "ref");
  },
  fromFirestore: function (
    snapshot: firebase.firestore.QueryDocumentSnapshot
  ): MyModel {
    const model = snapshot.data();
    // Some kind of validation, etc.
    model.ref = snapshot.ref;
    return model as MyModel;
  },
};

const modelDocumentRef = firebase.firestore()
  .collection("/models")
  .doc("some_id")
  .withConverter(modelConverter);

  modelDocumentRef.onSnapshot((snapshot) => {
  const model = snapshot.data();
  const modelSubcollectionRef = model?.ref.collection("/subcollection");
});
```

Runtime output:

```
Uncaught (in promise) TypeError: model.collection is not a function
```

Instead I have to do:

```
  fromFirestore: function (
    snapshot: firebase.firestore.QueryDocumentSnapshot
  ): MyModel {
    const model = snapshot.data();
    // Some kind of validation, etc.
>   model.ref = firebase.firestore().doc(snapshot.ref.path);
    return model as MyModel;
  },

```

**Comment by dconeybe** [comment]
This bug looks eerily similar to https://github.com/firebase/firebase-js-sdk/issues/4125, which was fixed by https://github.com/firebase/firebase-js-sdk/pull/4136 in 8.1.2 (December 3, 2020). I'll investigate and reply back.

**Comment by dconeybe** [comment]
I've confirmed this issue and it is indeed a bug. I'm working on a fix. In the meantime, your workaround makes perfect sense.

**Comment by letsgamedotdev** [comment]
> I've confirmed this issue and it is indeed a bug. I'm working on a fix. In the meantime, your workaround makes perfect sense.

This also happens with admin.firestore by the way. Not sure if the projects are related or if I should make another issue ticket.

**Comment by dconeybe** [comment]
If this bug is also happening with the admin sdk then it can't possibly be the same root cause. Please open a new ticket for it.

**Comment by dconeybe** [comment]
FYI the fix for this bug was released in version 8.2.4 on January 21, 2021: https://firebase.google.com/support/release-notes/js#version_824_-_january_21_2021
